### PR TITLE
controller: old pods should block deployment completeness

### PR DESCRIPTION
**Pull request by 0xmichalis**
Fixes https://github.com/kubernetes/kubernetes/issues/40207

### CODE REVIEW

**Comment by kargakis** [comment]
@janetkuo ptal, we need to cherry-pick this back to 1.5

File: pkg/controller/deployment/util/deployment_util.go [hunk 1]
```diff
@@ -816,9 +816,10 @@ func IsRollingUpdate(deployment *extensions.Deployment) bool {
 }
 
 // DeploymentComplete considers a deployment to be complete once its desired replicas equals its
```
Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go]: Would you add a sentence that summarizes what "complete" means? Something like:
```go
// DeploymentComplete checks if a deployment is complete, i.e. the deployment has been
// rolled out successfully. A deployment is complete if:
// 1. its desired replicas..., and 
// 2. ..
```

Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go]: We'll need to modify docs https://kubernetes.io/docs/user-guide/deployments/#complete-deployment

Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go]: Do we expose `Complete` state to the users?

Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: > Would you add a sentence that summarizes what "complete" means? Something like:

That's what the comment is about already, isn't it clear?

> We'll need to modify docs https://kubernetes.io/docs/user-guide/deployments/#complete-deployment

Yes, I will open a PR there.

> Do we expose Complete state to the users?

Not sure what you mean. We expose a Progressing=True condition with NewReplicaSetAvailable as a reason once a deployment completes.

Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go]: > We expose a Progressing=True condition with NewReplicaSetAvailable as a reason once a deployment completes.

I meant that we should also mention this in https://kubernetes.io/docs/user-guide/deployments/#complete-deployment (right now it's documented under failed deployment which makes it less obvious)

**Review by janetkuo — APPROVED** [review]

**Comment by kargakis** [comment]
@mikedanese @janetkuo can one of you guys approve this? Thanks

**Comment by spxtr** [comment]
@k8s-bot bazel test this
#40526

**Pull request merged**